### PR TITLE
[Bugfix] Point OnnxGz files to `model.onnx` but download and extract `model.onnx.tar.gz`

### PR DESCRIPTION
**Pull request by rahul-tuli**
Big LLMS are generally saved with external data; this external data along with model.onnx live inside `model.onnx.tar.gz` tarball on sparsezoo; recently we added support to download the tarfile and emulate a normal onnx model file using a wrapper class `OnnxGz`; when adding this support a bug was introduced where `model.onnx` was present twice in the model path; leading to errors while creating ort sessions as 

### TEST PLAN:

- [X] Manual inspection using attached script
- [X] Add automated tests to capture conditions where extraction was failing
- [X] Test with deepsparse main

Test Script:
```
from sparsezoo import Model
from pathlib import Path


def local_test(stub: str):
    model = Model(stub)
    onnx_model = model.onnx_model
    model_path = onnx_model.path
    onnx_model_path = Path(model_path)
    assert onnx_model_path.exists(), "Path to onnx model does not exist"
    assert (count := str(onnx_model_path).count("model.onnx")) == 1, f"Path to onnx model contains model.onnx {count} times"

    # create an ORT session with the model
    import onnxruntime as ort
    session = ort.InferenceSession(str(onnx_model_path), providers=["CPUExecutionProvider"])
    assert session is not None, "ORT session is None"

if __name__ == '__main__':
    # stub = "zoo:nlg/text_generation/opt-13b/pytorch/huggingface/opt_pretrain/pruned50_quantW8A8-none"
    stub = "zoo:cv/classification/mobilenet_v1-1.0/pytorch/sparseml/imagenet/pruned_quant-moderate"
    local_test(stub=stub)
```

Before this PR: The onnx_model_path had an extra model.onnx in it's path leading to errors like:
https://github.com/neuralmagic/deepsparse/actions/runs/5869604261/job/15914935604?pr=1150#step:7:138

After this PR the test script runs successfully, and the base test errors on deepsparse side are green

---
- To see the specific tasks where the Asana app for GitHub is being used, see below:
  - https://app.asana.com/0/0/1205276886236935

### CODE REVIEW

File: src/sparsezoo/model/model.py [hunk 1]
```diff
@@ -157,7 +157,7 @@ def __init__(self, source: str, download_path: Optional[str] = None):
             stub_params=self.stub_params,
         )
 
-        self._onnx_gz: Directory = self._directory_from_files(
+        self._onnx_gz: OnnxGz = self._directory_from_files(
```
Comment (rahul-tuli) [inline src/sparsezoo/model/model.py:160]: Changes in this file are merely cosmetic and could be a part of a separate PR

File: src/sparsezoo/model/model.py [hunk 2]
```diff
@@ -587,12 +587,12 @@ def _directory_from_files(
         regex: Optional[bool] = False,
         allow_multiple_outputs: Optional[bool] = False,
         **kwargs: object,
-    ) -> Union[Directory, None]:
+    ) -> Union[List[Union[Directory, Any, None]], List[Directory], None]:
```
Comment (rahul-tuli) [inline src/sparsezoo/model/model.py:590]: Fix typing

File: src/sparsezoo/model/utils.py [hunk 3]
```diff
@@ -576,11 +577,21 @@ def _copy_file_contents(
                 copy_path = os.path.join(output_dir, os.path.basename(_file.path))
                 _copy_and_overwrite(_file.path, copy_path, shutil.copyfile)
         elif isinstance(file, OnnxGz):
-            # copy all contents of unzipped onnx.tar.gz file to top level of output
-            onnx_gz_path = (
-                os.path.dirname(file.path) if os.path.isfile(file.path) else file.path
-            )
-            shutil.copytree(onnx_gz_path, output_dir, dirs_exist_ok=True)
+            # download and unzip model.onnx.tar.gz
```
Comment (rahul-tuli) [inline src/sparsezoo/model/utils.py:580]: copy contents of OnnxGz file based on tarfile members rather than copying the whole parent directory

File: src/sparsezoo/objects/directories.py [hunk 4]
```diff
@@ -283,14 +284,90 @@ class OnnxGz(Directory):
...
+    Class Invariants:
+        - `self.name` attribute of this class will point to the name of the onnx file
+        - `self._path` and `self.path` will point to the path of the extracted
+            onnx model
     """
 
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+
+    def __repr__(self):
```
Comment (rahul-tuli) [inline src/sparsezoo/objects/directories.py:297]: Add __repr__ for easier debugging

File: src/sparsezoo/objects/directories.py [hunk 5]
```diff
@@ -283,14 +284,90 @@ class OnnxGz(Directory):
...
+            raise FileNotFoundError(
+                f"Expected to find model.onnx at {onnx_model_path}, "
+                "but it does not exist."
+            )
+
+        # point _path to model.onnx
+        self._path = str(onnx_model_path)
         return self._path
+
+    def _check_if_extracted(self):
```
Comment (rahul-tuli) [inline src/sparsezoo/objects/directories.py:348]: checks if model.onnx.tar.gz was extracted based on all members inside the tarball, accordingly sets `is_archive`

File: src/sparsezoo/objects/directory.py [hunk 6]
```diff
@@ -284,9 +284,15 @@ def unzip(self, extract_directory: Optional[str] = None, force: bool = False):
...
-
-        if not os.path.exists(path) or force:  # do not re-unzip if not forced
+        is_model_gz = name == "model.onnx"
+        path = (
+            extract_directory if is_model_gz else os.path.join(extract_directory, name)
+        )
+        # if is_model_gz then path would point to the tarfile(s)
+        # parent directory and must be unzipped irrespective of existence
+        if (
+            is_model_gz or not os.path.exists(path) or force
```
Comment (rahul-tuli) [inline src/sparsezoo/objects/directory.py:294]: extract model.onnx.tar.gz

File: src/sparsezoo/objects/directory.py [hunk 7]
```diff
@@ -340,7 +346,7 @@ def is_directory(file: File) -> bool:
 def _possibly_convert_files_to_directories(files: List[File]) -> List[File]:
     return [
         Directory.from_file(file)
-        if (is_directory(file) and not isinstance(file, Directory))
+        if not isinstance(file, Directory) and is_directory(file)
```
Comment (rahul-tuli) [inline src/sparsezoo/objects/directory.py:349]: Move isinstance check first to avoid calling path property on Directory objects

File: src/sparsezoo/model/model.py (hunk 2 above)
Comment (dsikka) [reply src/sparsezoo/model/model.py:590]: `Any` doesn't cover the `None` case?

File: src/sparsezoo/model/utils.py [hunk 8]
```diff
@@ -576,11 +577,21 @@ def _copy_file_contents(
...
+            #  top level of output
+            for name in tarfile.open(model_gz_path).getnames():
+                unzipped_file_path = model_gz_path.with_name(name)
+                copy_location = Path(output_dir) / name
+                _copy_and_overwrite(
+                    str(unzipped_file_path), str(copy_location), shutil.copyfile
+                )
+            # copy the model.onnx.tar.gz file to the top level of output
+            tar_copy_path = Path(output_dir) / model_gz_path.name
+            _copy_and_overwrite(str(model_gz_path), str(tar_copy_path), shutil.copyfile)
```
Comment (dsikka) [inline src/sparsezoo/model/utils.py:593]: why do you need to copy the tar file after already copying its untarred content in the for loop above?

File: src/sparsezoo/objects/directories.py [hunk 9]
```diff
@@ -283,14 +284,90 @@ class OnnxGz(Directory):
...
+    def __repr__(self):
+        return f"OnnxGz(name={self.name}, path={self._path})"
+
     @property
     def path(self):
-        _ = super().path  # call self.path to download initial file if not already
+        """
+        Assumptions:
+            - the tarball contains only one onnx model, with the
+                name `model.onnx`
```
Comment (Satrat) [inline src/sparsezoo/objects/directories.py:305]: This is not the case for the codegen models on sparsezoo, for at least one of them the tarball contains "model_nocache.onnx" Could we just search the directory for a .onnx file instead? We would still be assuming there is only one onnx model file in the folder

File: src/sparsezoo/model/model.py (hunk 2 above)
Comment (rahul-tuli) [reply src/sparsezoo/model/model.py:590]: Hmm, you are correct it should

File: src/sparsezoo/objects/directories.py [hunk 10]
```diff
@@ -283,14 +284,90 @@ class OnnxGz(Directory):
...
+
+        self._check_if_extracted()
         if self.is_archive:
+            # if the tarball is not extracted, extract it
             self.unzip()
-        if os.path.isdir(self._path) and "model.onnx" in os.listdir(self._path):
-            # if unzipped into a directory, refer directly to model.onnx
-            self._path = os.path.join(self._path, "model.onnx")
+
+        onnx_model_path = expected_path.with_name(name="model.onnx")
```
Comment (Satrat) [inline src/sparsezoo/objects/directories.py:337]: Do we need to enforce the model name as `model.onnx`? I think just searching for the file extension would be more flexible

File: src/sparsezoo/objects/directory.py [hunk 11]
```diff
@@ -284,9 +284,15 @@ def unzip(self, extract_directory: Optional[str] = None, force: bool = False):
             )
 
         name = ".".join(self.name.split(".")[:-2])
-        path = os.path.join(extract_directory, name)
-
-        if not os.path.exists(path) or force:  # do not re-unzip if not forced
+        is_model_gz = name == "model.onnx"
```
Comment (Satrat) [inline src/sparsezoo/objects/directory.py:287]: See other comments, not sure about hard coding this

File: src/sparsezoo/objects/directories.py [hunk 12]
```diff
@@ -283,14 +284,90 @@ class OnnxGz(Directory):
...
+            expected_path = expected_path / "model.onnx.tar.gz"
+
+        model_gz_path = expected_path.with_name(name="model.onnx.tar.gz")
+        # assert all members of  model.onnx.tar.gz have been extracted
+        for zipped_filename in tarfile.open(model_gz_path).getnames():
+            unzipped_file_path = expected_path.with_name(zipped_filename)
+            if not unzipped_file_path.exists():
+                _LOGGER.info(f"{unzipped_file_path} does not exist, was it extracted?")
+                self.is_archive = True
+                return
```
Comment (Satrat) [inline src/sparsezoo/objects/directories.py:372]: This is duplicated in another function, could we break it out into a helper?

**Review by Satrat — CHANGES_REQUESTED** [review]
I'm pretty sure the file we want to unzip is model.tar.gz and not model.onnx.tar.gz if the purpose of this PR is to extract the model.data file.

[The 2B codegen stub](https://sparsezoo.neuralmagic.com/models/codegen_mono-2b-bigpython_bigquery_thepile-base?hardware=deepsparse-c6i.12xlarge&tab=4) shows model.tar.gz as 6.4GB and model.onnx.tar.gz as only 594KB.

In general, I think it would be a good idea to upload a small model to sparsezoo that has its data stored externally so we can catch the issue in the test suite without having to run giant models

File: src/sparsezoo/model/utils.py (hunk 8 above)
Comment (rahul-tuli) [reply src/sparsezoo/model/utils.py:593]: There's no need per se; we can avoid copying over the tarball; from how I see it, it doesn't hurt or give us anything extra. I kept it in case someone wants to manually check it's contents, but I don't have a strong argument for either case.

File: src/sparsezoo/objects/directories.py (hunk 9 above)
Comment (rahul-tuli) [reply src/sparsezoo/objects/directories.py:305]: This is something we want to enforce with `model.onnx.tar.gz` (that it always contains a `model.onnx` along with external data files if any) I could see a future where we support other model names, but they would be handled separately and not via `model.onnx.tar.gz`, the codegen 2B model you mentioned has to be updated on sparsezoo

File: src/sparsezoo/objects/directories.py (hunk 10 above)
Comment (rahul-tuli) [reply src/sparsezoo/objects/directories.py:337]: I agree it would be more flexible, but let's leave this as is for now as its the only check enforcing constraints on the contents of `model.onnx.tar.gz` we can update this if the constraints change?

File: src/sparsezoo/objects/directory.py (hunk 11 above)
Comment (rahul-tuli) [reply src/sparsezoo/objects/directory.py:287]: This would work for now, as we are enforcing *decided* constraints on the contents of `model.onnx.tar.gz` but, am open to changing this logic if the constraints are ever relaxed

File: src/sparsezoo/objects/directories.py (hunk 12 above)
Comment (rahul-tuli) [reply src/sparsezoo/objects/directories.py:372]: That method uses this same iteration logic to copy/overwrite extracted files; only the iteration part is common and imo breaking the iteration part out might not be as beneficial. Thoughts? Happy to break this out if this is a hard blocker

**Comment by rahul-tuli** [comment]
> I'm pretty sure the file we want to unzip is model.tar.gz and not model.onnx.tar.gz if the purpose of this PR is to extract the model.data file.

making a comment here that the linked model needs to be updated on sparsezoo; and the onnx model `model.onnx` along with external_data if any should be inside `model.onnx.tar.gz`

[The 2B codegen stub](https://sparsezoo.neuralmagic.com/models/codegen_mono-2b-bigpython_bigquery_thepile-base?hardware=deepsparse-c6i.12xlarge&tab=4) shows model.tar.gz as 6.4GB and model.onnx.tar.gz as only 594KB.

In general, I think it would be a good idea to upload a small model to sparsezoo that has its data stored externally so we can catch the issue in the test suite without having to run giant models

File: src/sparsezoo/objects/directories.py (hunk 12 above)
Comment (Satrat) [reply src/sparsezoo/objects/directories.py:372]: Makes sense, I'm all good to leave it

File: src/sparsezoo/model/utils.py (hunk 8 above)
Comment (Satrat) [reply src/sparsezoo/model/utils.py:593]: My vote would be to avoid copying it, for large models especially we don't want to keep around unnecessary files

File: src/sparsezoo/model/model.py (hunk 2 above)
Comment (rahul-tuli) [reply src/sparsezoo/model/model.py:590]: But I believe it's nicer to make it explicit, will remove if it's a hard blocker for you

**Review by Satrat — APPROVED** [review]

File: src/sparsezoo/model/utils.py (hunk 8 above)
Comment (rahul-tuli) [reply src/sparsezoo/model/utils.py:593]: Lol this function is only used in tests so I guess either way is fine

File: src/sparsezoo/objects/directories.py (hunk 9 above)
Comment (dsikka) [reply src/sparsezoo/objects/directories.py:305]: the other case where this may change quickly is for any pipeline requiring more than one model (encoder/decoder pipelines). We haven't come up with a definite plan for these on sparsezoo but I think *.onnx would be more generic?

File: src/sparsezoo/objects/directories.py (hunk 10 above)
Comment (dsikka) [reply src/sparsezoo/objects/directories.py:337]: you could do something where you check if multiple *.onnx files are present. If not, ensure the single one present is `model.onnx`

**Review by dsikka — COMMENTED** [review]
Two comments. otherwise LGTM

File: src/sparsezoo/objects/directories.py (hunk 9 above)
Comment (rahul-tuli) [reply src/sparsezoo/objects/directories.py:305]: Let's leave it until we have a definite plan for these on sparsezoo; don't want to make any decisions here that are bound to change

File: src/sparsezoo/objects/directories.py (hunk 10 above)
Comment (rahul-tuli) [reply src/sparsezoo/objects/directories.py:337]: We would still need a definite plan on which of those onnx models should `Model(stub).onnx_model` point to; so I think this update would make more sense there?

File: src/sparsezoo/objects/directory.py (hunk 6 above)
Comment (bfineran) [inline src/sparsezoo/objects/directory.py:294]: why do we always unzip if `is_model_gz`? what would happen if the tarball still exists but has already been unzipped?

File: src/sparsezoo/objects/directories.py [hunk 13]
```diff
@@ -283,14 +284,90 @@ class OnnxGz(Directory):
...
+        if expected_path.is_dir():
+            # point expected to model.onnx.tar.gz
+            expected_path = expected_path / "model.onnx.tar.gz"
+
+        model_gz_path = expected_path.with_name(name="model.onnx.tar.gz")
+        # assert all members of  model.onnx.tar.gz have been extracted
+        for zipped_filename in tarfile.open(model_gz_path).getnames():
+            unzipped_file_path = expected_path.with_name(zipped_filename)
+            if not unzipped_file_path.exists():
+                _LOGGER.info(f"{unzipped_file_path} does not exist, was it extracted?")
```
Comment (bfineran) [inline src/sparsezoo/objects/directories.py:370]: should remove log - user can't do anything with this. potentially move to debug since this would be a weird state

File: src/sparsezoo/objects/directory.py (hunk 6 above)
Comment (rahul-tuli) [reply src/sparsezoo/objects/directory.py:294]: synced offline!

File: src/sparsezoo/objects/directories.py (hunk 9 above)
Comment (rahul-tuli) [reply src/sparsezoo/objects/directories.py:305]: Done!

**Review by bfineran — APPROVED** [review]
